subprojects/common/core/src/main/java/hu/bme/mit/theta/core/utils/ExprSimplifier.java

```diff
@@ -412,7 +414,13 @@ private ExprSimplifier() {
 		Expr<ArrayType<IT, ET>> arr = simplify(expr.getArray(), val);
 		Expr<IT> index = simplify(expr.getIndex(), val);
 		if (arr instanceof LitExpr<?> && index instanceof LitExpr<?>) {
```

Comment:
I would add a comment here why the condition `index instanceof LitExpr<?>` is required. If I'm not mistaken it is required because for literals `equals` is precise. However, if we would have `x+1` as an index and we don't know `x`, then we cannot really decide whether to return some element with a key, or a default element. Like if your array has `1 => 1, 2 => 2, else => 3` then `arr[x+1]` can basically correspond to any element, including the `else` as well.

Reply:
I think a comment like the following would do:

> The index is required to be a literal so that we can use 'equals' to compare it against existing keys in the array